File: src/core/camera.ts
import { Point, Size } from './types';

export class Camera {
  public offsetX: number = 0;
  public offsetY: number = 0;
  public cellSize: number = 16;

  public static readonly MinCellSize: number = 2;
  public static readonly MaxCellSize: number = 64;

  public static readonly MinY: number = -159;
  public static readonly MaxY: number = 159;

  public screenToGrid(screen: Point): Point {
    return {
      x: Math.floor((screen.x - this.offsetX) / this.cellSize),
      y: Math.floor((screen.y - this.offsetY) / this.cellSize),
    };
  }

  public gridToScreen(gx: number, gy: number): Point {
    return {
      x: gx * this.cellSize + this.offsetX,
      y: gy * this.cellSize + this.offsetY,
    };
  }

  public getVisibleGridRect(clientSize: Size): {
    left: number;
    top: number;
    right: number;
    bottom: number;
  } {
    const left = Math.floor(-this.offsetX / this.cellSize) - 1;
    const top = Math.floor(-this.offsetY / this.cellSize) - 1;
    const right = Math.ceil((clientSize.width - this.offsetX) / this.cellSize) + 1;
    const bottom = Math.ceil((clientSize.height - this.offsetY) / this.cellSize) + 1;

    return { left, top, right, bottom };
  }

  public zoomAt(cursor: Point, delta: number): void {
    const gridXBefore = (cursor.x - this.offsetX) / this.cellSize;
    const gridYBefore = (cursor.y - this.offsetY) / this.cellSize;

    const step = delta > 0 ? 1 : -1;
    if (this.cellSize < 8) {
      this.cellSize += step;
    } else if (this.cellSize < 16) {
      this.cellSize += step * 2;
    } else {
      this.cellSize += step * 4;
    }

    this.cellSize = Math.min(
      Math.max(this.cellSize, Camera.MinCellSize),
      Camera.MaxCellSize
    );

    this.offsetX = cursor.x - gridXBefore * this.cellSize;
    this.offsetY = cursor.y - gridYBefore * this.cellSize;
  }

  public centerOn(clientSize: Size, gridX: number = 0, gridY: number = 0): void {
    this.offsetX = clientSize.width / 2 - gridX * this.cellSize;
    this.offsetY = clientSize.height / 2 - gridY * this.cellSize;
  }
}

File: Canvas/Camera.cs
namespace MCBp.Canvas;

public class Camera
{
    public float OffsetX;
    public float OffsetY;
    public int CellSize = 16;

    public const int MinCellSize = 2;
    public const int MaxCellSize = 64;

    public const int MinY = -159;
    public const int MaxY = 159;

    public Point ScreenToGrid(Point screen)
    {
        int gx = (int)Math.Floor((screen.X - OffsetX) / CellSize);
        int gy = (int)Math.Floor((screen.Y - OffsetY) / CellSize);
        return new Point(gx, gy);
    }

    public PointF GridToScreen(int gx, int gy)
    {
        return new PointF(
            gx * CellSize + OffsetX,
            gy * CellSize + OffsetY);
    }

    public Rectangle GetVisibleGridRect(Size clientSize)
    {
        int left = (int)Math.Floor(-OffsetX / CellSize) - 1;
        int top = (int)Math.Floor(-OffsetY / CellSize) - 1;
        int right = (int)Math.Ceiling((clientSize.Width - OffsetX) / CellSize) + 1;
        int bottom = (int)Math.Ceiling((clientSize.Height - OffsetY) / CellSize) + 1;

        return new Rectangle(left, top, right - left, bottom - top);
    }

    public void ZoomAt(Point cursor, int delta)
    {
        float gridXBefore = (cursor.X - OffsetX) / CellSize;
        float gridYBefore = (cursor.Y - OffsetY) / CellSize;

        int step = delta > 0 ? 1 : -1;
        if (CellSize < 8)
            CellSize += step;
        else if (CellSize < 16)
            CellSize += step * 2;
        else
            CellSize += step * 4;

        CellSize = Math.Clamp(CellSize, MinCellSize, MaxCellSize);

        OffsetX = cursor.X - gridXBefore * CellSize;
        OffsetY = cursor.Y - gridYBefore * CellSize;
    }

    public void CenterOn(Size clientSize, int gridX = 0, int gridY = 0)
    {
        OffsetX = clientSize.Width / 2f - gridX * CellSize;
        OffsetY = clientSize.Height / 2f - gridY * CellSize;
    }
}
